zetaclient/chains/solana/signer/signer.go

```diff
@@ -0,0 +1,280 @@
+	baseSigner := base.NewSigner(chain, tss, ts, logger)
+
+	// parse gateway ID and PDA
+	gatewayID, pda, err := contract.ParseGatewayIDAndPda(chainParams.GatewayAddress)
+	if err != nil {
+		return nil, errors.Wrapf(err, "cannot parse gateway address %s", chainParams.GatewayAddress)
+	}
+	logger.Std.Info().Msgf("Solana fee payer address: %s", solanaKey.PublicKey())
+
+	// create solana observer
+	return &Signer{
+		Signer:            baseSigner,
+		client:            solClient,
+		solanaFeePayerKey: solanaKey,
+		gatewayID:         gatewayID,
+		pda:               pda,
+	}, nil
+}
+
+// SignMsgWithdraw signs a message for Solana gateway withdraw/withdraw_spl instruction
+func (signer *Signer) SignMsgWithdraw(
+	ctx context.Context,
+	params *types.OutboundParams,
+	height uint64,
+) (*contract.MsgWithdraw, error) {
+	chain := signer.Chain()
+	// #nosec G115 always positive
+	chainID := uint64(signer.Chain().ChainId)
+	nonce := params.TssNonce
+	amount := params.Amount.Uint64()
+
+	// check receiver address
+	to, err := chains.DecodeSolanaWalletAddress(params.Receiver)
+	if err != nil {
+		return nil, errors.Wrapf(err, "cannot decode receiver address %s", params.Receiver)
+	}
+
+	// prepare withdraw msg and compute hash
+	msg := contract.NewMsgWithdraw(chainID, nonce, amount, to)
+	msgHash := msg.Hash()
+
+	// sign the message with TSS to get an ECDSA signature.
+	// the produced signature is in the [R || S || V] format where V is 0 or 1.
+	signature, err := signer.TSS().Sign(ctx, msgHash[:], height, nonce, chain.ChainId, "")
+	if err != nil {
+		return nil, errors.Wrap(err, "Key-sign failed")
+	}
+	signer.Logger().Std.Info().Msgf("Key-sign succeed for chain %d nonce %d", chainID, nonce)
+
+	// attach the signature and return
+	return msg.WithSignature(signature), nil
+}
+
+// SignWithdrawTx signs the Solana gateway 'withdraw' transaction specified by 'msg'
+func (signer *Signer) SignWithdrawTx(ctx context.Context, msg contract.MsgWithdraw) (*solana.Transaction, error) {
+	// create withdraw instruction with program call data
+	var err error
+	var inst solana.GenericInstruction
+	inst.DataBytes, err = borsh.Serialize(contract.WithdrawInstructionParams{
+		Discriminator: contract.DiscriminatorWithdraw(),
+		Amount:        msg.Amount,
+		Signature:     msg.SigRS(),
+		RecoveryID:    msg.SigV(),
+		MessageHash:   msg.Hash(),
+		Nonce:         msg.Nonce,
+	})
+	if err != nil {
+		return nil, errors.Wrap(err, "cannot serialize withdraw instruction")
+	}
+
+	// attach required accounts to the instruction
+	privkey := signer.solanaFeePayerKey
+	var accountSlice []*solana.AccountMeta
+	accountSlice = append(accountSlice, solana.Meta(privkey.PublicKey()).WRITE().SIGNER())
+	accountSlice = append(accountSlice, solana.Meta(signer.pda).WRITE())
+	accountSlice = append(accountSlice, solana.Meta(msg.To).WRITE())
+	accountSlice = append(accountSlice, solana.Meta(signer.gatewayID))
+	inst.ProgID = signer.gatewayID
+	inst.AccountValues = accountSlice
+
```

Comment:
Seems we could have an helper function for this

Reply:
https://github.com/zeta-chain/node/pull/2560/commits/ada365437e0782d6d2f6725ca87d6c53a687118d